util/file_reader_writer.h

```diff
@@ -70,24 +94,27 @@ class RandomAccessFileReader {
   HistogramImpl*  file_read_hist_;
   RateLimiter* rate_limiter_;
   bool for_compaction_;
+  std::vector<std::shared_ptr<EventListener>> listeners_;
+  bool enable_listeners_;
 
  public:
-  explicit RandomAccessFileReader(std::unique_ptr<RandomAccessFile>&& raf,
-                                  std::string _file_name,
-                                  Env* env = nullptr,
-                                  Statistics* stats = nullptr,
-                                  uint32_t hist_type = 0,
-                                  HistogramImpl* file_read_hist = nullptr,
-                                  RateLimiter* rate_limiter = nullptr,
-                                  bool for_compaction = false)
+  explicit RandomAccessFileReader(
+      std::unique_ptr<RandomAccessFile>&& raf, std::string _file_name,
+      Env* env = nullptr, Statistics* stats = nullptr, uint32_t hist_type = 0,
+      HistogramImpl* file_read_hist = nullptr,
+      RateLimiter* rate_limiter = nullptr, bool for_compaction = false,
+      const std::vector<std::shared_ptr<EventListener>>& listeners = {},
+      bool enable_listeners = false)
```

Comment:
Where will this be enabled? Asking as I don't see enable_listeners to be set to true anywhere. Same question for WritableFileWriter too.

Reply:
This will be enabled by an option so that users can control whether they would like to enable reporting of each individual file IO and ready to tolerate the possible performance overhead. The option is not implemented yet and will involve changing every invocation of constructor. Therefore, I think it's better to do that in a separate PR.

Reply:
I see, thanks.
- Can you add this information (and the future plan ) to the PR summary so that everyone is aware? 
- Since we are trying to move away from introducing new options as much as we can, is there some other way to control these new listeners when needed? 
- I believe the code to enable these listeners should also go along with this PR. I'll leave it up to you though. If no, may be introducing enable_listeners param should also be moved to that new PR.

Reply:
Btw, I am super interested in this PR, and hence I was taking a deeper look :)

Reply:
> I see, thanks.
> 
> * Can you add this information (and the future plan ) to the PR summary so that everyone is aware?
Sure, will update PR description as well as code comments.
> * Since we are trying to move away from introducing new options as much as we can, is there some other way to control these new listeners when needed?
I don't have a good solution to this so far. The user has to convey this information to RocksDB, and options seem to be the only place. Existing options do not seem to suffice.
> * I believe the code to enable these listeners should also go along with this PR. I'll leave it up to you though. If not, may be introducing enable_listeners param should also be moved to that new PR.
I tend to agree...there are so many calls of `new WritableFileWriter` and/or `new RandomAccessFileReader`. I don't want to use default argument, and enabling this feature requires changing the argument lists of these calls. Let me think about it more.

Reply:
@sagar0 I updated description with test plan. A new db option does not seem necessary if we can add a new function `ShouldBeNotifiedOnFileIO` to `EventListener`. This PR also enables the feature in the situations that we are interested in.